### ISSUE TITLE
[Lesson] Building on the Data Preparation Lesson

### ISSUE THREAD

**Issue by paladique** [microsoft/Data-Science-For-Beginners#140]
The [data preparation lesson](https://github.com/microsoft/Data-Science-For-Beginners/tree/main/2-Working-With-Data/08-data-preparation is looking pretty small, some additions to this lesson would make this spectacular.

**Comment by IndraP24** [comment]
I would like to add some more content to the lesson if it is available for us to contribute to.

**Comment by jlooper** [comment]
hi @IndraP24 if you'd like to propose a PR we will take a look! thank you

**Comment by IndraP24** [comment]
Sure @jlooper. Will propose one as soon as I'm done with adding the contents.

**Comment by nirmalya8** [comment]
Hey, Can I add some more code blocks and elaborate on some of the explanations in the [notebooks.ipynb](https://github.com/microsoft/Data-Science-For-Beginners/blob/main/2-Working-With-Data/08-data-preparation/notebook.ipynb) file under data preparation? Can I work on it and propose a PR?

**Comment by jlooper** [comment]
> Hey, Can I add some more code blocks and elaborate on some of the explanations in the [notebooks.ipynb](https://github.com/microsoft/Data-Science-For-Beginners/blob/main/2-Working-With-Data/08-data-preparation/notebook.ipynb) file under data preparation? Can I work on it and propose a PR?

Yes, but I’d recommend working with other contributors to make a good lesson. Maybe outline here your proposed additions

**Comment by nirmalya8** [comment]
@jlooper 
I want to implement the following:
- [ ] Code examples
    - [ ]  Shape of the dataframe
    - [ ]  Dataframe.describe()
    - [ ]  Ordinal Encoding
    - [ ]  One Hot Encoding
    - [ ]  Imputations using sklearn's SimpleImputer
- [ ]  Explanations
    - [ ]  When do we drop rows with missing values?
    - [ ]  When do we replace missing values?
    - [ ]  For numerical data, when do replace missing values with mean and when do we do so with the median?
    - [ ]  Explanation for Encoding: Both Ordinal and One Hot

Will this be okay? Should I go forward and prepare the notebook and make a Pull Request?

**Comment by IndraP24** [comment]
That's great @nirmalya8. It doesn't conflict with anything I'm adding as a part of the README explanations.

I'm creating a new Section: Common Data Preparation Tasks with explanations of the following:
- [x] Data Cleaning
- [x] Feature Selection
- [ ] Data Transforms
- [ ] Feature Engineering
- [ ] Dimensionality Reduction

I'm only adding brief explanations as of now.

**Comment by jlooper** [comment]
hi, we'll need to do two things: prepare a solution notebook and prepare the .md explanation lesson itself. I recommend avoiding using Sklearn right now, as we use it so much in the ML curriculum (aka.ms/ml-beginners). Let's see how far we can get using pandas and numpy.

Let's also create the lesson as a 'backwards designed' lesson. First, list the concepts you want to drive home:

- When do we drop rows with missing values?
- When do we replace missing values?
- For numerical data, when do replace missing values with mean and when do we do so with the median?
- For textual data, how can you manage corrupt data, dates, and the like
- Why worry about encoding? When do we need to use it

Then, for each of those sections, include a code sample based on a small dataset that exhibits the problem you are trying to correct and show how to correct these problems.

Does that sound good?

**Comment by nirmalya8** [comment]
Yeah, @jlooper by explanations I meant this only. So, basically, whatever I listed, it would be better to create a new notebook for it(the backwards-designed thing you mentioned). Stating the reasons why what is done and then showing it via some code blocks. So, it isn't necessary for me to edit the existing notebook. And as for the readme, @IndraP24 is doing that. Is that okay?

**Comment by jlooper** [comment]
hi, we want just one solution notebook, so the new content needs to dovetail to the old content. It's also important that the README reflect exactly what's in that solution notebook; you can see how it's done in the visualization lessons. Do you think you can create this same type of experience with this new content?

**Comment by jlooper** [comment]
Looking at the lesson structure, I see what's going on. Most of the content is in the base notebook. I think it will be better to restructure the lesson by adding to the README and the base notebook, to add the missing pieces. Then, I will move this full notebook over to the 'solution' folder and create a root blank notebook for students to use. 

So the work order is: 

- enhance the current README to reflect what's in the current base notebook with explanations
- enhance the lesson itself in the README with some further content and explanations
- enhance the base notebook to reflect the solutions explained in the README
- move that base notebook to a /solution folder and create a blank notebook in the root

Sorry for confusion; this is a case of a little conflicting curriculum design concept

**Comment by nirmalya8** [comment]
> hi, we want just one solution notebook, so the new content needs to dovetail to the old content. It's also important that the README reflect exactly what's in that solution notebook; you can see how it's done in the visualization lessons. Do you think you can create this same type of experience with this new content?

Okay, I get what you are saying. As the [solutions notebook](https://github.com/microsoft/Data-Science-For-Beginners/blob/main/2-Working-With-Data/08-data-preparation/notebook.ipynb) is already there, and it contains some stuff, I would just edit that and fit in the things I mentioned in that notebook. As for the readme,  I won't go beyond what I mentioned in [this comment](https://github.com/microsoft/Data-Science-For-Beginners/issues/140#issuecomment-932979978). So, I am starting with it. @jlooper

**Comment by IndraP24** [comment]
> Looking at the lesson structure, I see what's going on. Most of the content is in the base notebook. I think it will be better to restructure the lesson by adding to the README and the base notebook, to add the missing pieces. Then, I will move this full notebook over to the 'solution' folder and create a root blank notebook for students to use.
> 
> So the work order is:
> 
> * enhance the current README to reflect what's in the current base notebook with explanations
> * enhance the lesson itself in the README with some further content and explanations
> * enhance the base notebook to reflect the solutions explained in the README
> * move that base notebook to a /solution folder and create a blank notebook in the root
> 
> Sorry for confusion; this is a case of a little conflicting curriculum design concept

Alright, got it. I have already added some content to the README as per [this comment](https://github.com/microsoft/Data-Science-For-Beginners/issues/140#issuecomment-932987846).

I'll also go beyond that and  `enhance the current README to reflect what's in the current base notebook with explanations`. I'll follow the other lessons in this repo for the template and will stick with `pandas` and `numpy`.

**Comment by jlooper** [comment]
terrific! Thank you all!

**Comment by paladique** [comment]
What a wonderful discussion, thanks all!
